network/router_test.go

```diff
@@ -0,0 +1,365 @@
+	for i := 0; i < nbrRouter; i++ {
+		go func(j int) {
+			r := routers[j]
+			// expect nbrRouter - 1 messages
+			proc := newNSquareProc(t, r, nbrRouter-1, &wg2)
+			r.RegisterProcessor(proc, SimpleMessageType)
+			for k := 0; k < nbrRouter; k++ {
+				if k == j {
+					// don't send to yourself
+					continue
+				}
+				// send to everyone else
+				if err := r.Send(routers[k].id, &SimpleMessage{3}); err != nil {
+					t.Fatal(err)
+				}
+			}
+		}(i)
+	}
+	wg2.Wait()
+	for i := 0; i < nbrRouter; i++ {
+		r := routers[i]
+		if err := r.Stop(); err != nil {
+			t.Fatal(err)
+		}
+
+	}
+}
+
+// Test sending data back and forth using the sendSDAData
+func TestRouterSendMsgDuplexTCP(t *testing.T) {
+	testRouterSendMsgDuplex(t, NewTestRouterTCP)
+}
+
+func TestRouterSendMsgDuplexLocal(t *testing.T) {
+	testRouterSendMsgDuplex(t, NewTestRouterLocal)
+}
+func testRouterSendMsgDuplex(t *testing.T, fac routerFactory) {
+	h1, err1 := fac(2011)
+	h2, err2 := fac(2012)
+	if err1 != nil || err2 != nil {
+		t.Fatal("Could not setup hosts")
+	}
+	go h1.Start()
+	go h2.Start()
+
+	defer func() {
+		h1.Stop()
+		h2.Stop()
+		time.Sleep(250 * time.Millisecond)
+	}()
+
+	proc := &simpleMessageProc{t, make(chan SimpleMessage)}
+	h1.RegisterProcessor(proc, SimpleMessageType)
+	h2.RegisterProcessor(proc, SimpleMessageType)
+
+	msgSimple := &SimpleMessage{5}
+	err := h1.Send(h2.id, msgSimple)
+	if err != nil {
+		t.Fatal("Couldn't send message from h1 to h2", err)
+	}
+	msg := <-proc.relay
+	log.Lvl2("Received msg h1 -> h2", msg)
+
+	err = h2.Send(h1.id, msgSimple)
+	if err != nil {
+		t.Fatal("Couldn't send message from h2 to h1", err)
+	}
+	msg = <-proc.relay
+	log.Lvl2("Received msg h2 -> h1", msg)
+}
+
+func TestRouterExchange(t *testing.T) {
+	router1, err := NewTestRouterTCP(7878)
+	router2, err2 := NewTestRouterTCP(8787)
+	if err != nil || err2 != nil {
+		t.Fatal("Could not setup host", err, err2)
+	}
+
+	done := make(chan bool)
+	go func() {
+		done <- true
+		router1.Start()
+		done <- true
+	}()
+	<-done
+	// try correctly
+	c, err := NewTCPConn(router1.id.Address)
+	if err != nil {
+		t.Fatal("Couldn't connect to host1:", err)
+	}
+	if err := c.Send(router2.id); err != nil {
+		t.Fatal("Wrong negotiation")
+	}
+	// triggers the dispatching conditional branch error router.go:
+	//  `log.Lvl3("Error dispatching:", err)`
+	if err := router2.Send(router1.id, &SimpleMessage{12}); err != nil {
+		t.Fatal("Could not send")
+	}
+	c.Close()
+
+	// try messing with the connections here
+	c, err = NewTCPConn(router1.id.Address)
+	if err != nil {
+		t.Fatal("Couldn't connect to host1:", err)
+	}
+	// closing before sending
+	c.Close()
+	if err := c.Send(router2.id); err == nil {
+		t.Fatal("negotiation should have aborted")
+	}
+
+	// stop everything
+	log.Lvl4("Closing connections")
+	if err := router2.Stop(); err != nil {
+		t.Fatal("Couldn't close host", err)
+	}
+	if err := router1.Stop(); err != nil {
+		t.Fatal("Couldn't close host", err)
+	}
+	<-done
+}
```

Comment:
> as we cannot continue for most of the asserts we have

Let's not make big rules for that please because good units tests IMO should only need assert instead of require... but `require` is cool when we need it definitely !!